test/http/request.cpp

```diff
@@ -3,83 +3,204 @@
                  {"http/method", "GET"},
                  {"http/http_version", "HTTP/1.1"},
                  {"net/ca_bundle_path", "test/fixtures/certs.pem"}},
                 {
                     {"Accept", "*/*"},
                 },
-                "", [](Error error, Var<Response> response) {
-                    if (error != 0) {
-                        std::cout << "Error: " << (int)error << "\r\n";
-                        break_loop();
-                        return;
-                    }
-                    std::cout << "HTTP/" << response->http_major << "."
-                              << response->http_minor << " "
-                              << response->status_code << " " << response->reason
-                              << "\r\n";
-                    for (auto kv : response->headers) {
-                        std::cout << kv.first << ": " << kv.second << "\r\n";
-                    }
-                    std::cout << "\r\n";
-                    std::cout << response->body.substr(0, 128) << "\r\n";
-                    std::cout << "[snip]\r\n";
+                "",
+                [](Error error, Var<Response> response) {
+                    REQUIRE(!error);
+                    REQUIRE(response->status_code == 200);
+                    REQUIRE(md5(response->body) ==
+                            "1be9d96d157a3df328faa30e51faf63a");
                     break_loop();
                 });
     });
 }
 
+TEST_CASE("http::request() works as expected over Tor") {
+    loop_with_initial_event_and_connectivity([]() {
+        request(
+            {
+                {"http/url",
+                 "http://nexa.polito.it/nexafiles/ComeFunzionaInternet.pdf"},
+                {"http/method", "GET"},
+                {"http/http_version", "HTTP/1.1"},
+                {"Connection", "close"},
+                {"net/socks5_proxy", "127.0.0.1:9050"},
+            },
+            {
+                {"Accept", "*/*"},
+            },
+            "",
+            [&](Error error, Var<Response> response) {
+                REQUIRE(check_error_after_tor(error));
+                if (!error) {
+                    REQUIRE(response->status_code == 200);
+                    REQUIRE(md5(response->body) ==
+                            "efa2a8f1ba8a6335a8d696f91de69737");
+                }
+                break_loop();
+            });
+    });
+}
+
+TEST_CASE("http::request correctly receives errors") {
+    loop_with_initial_event_and_connectivity([]() {
+        request(
+            {
+                {"http/url", "http://nexa.polito.it:81/robots.txt"},
+                {"http/method", "GET"},
+                {"http/http_version", "HTTP/1.1"},
+                {"net/timeout", "3.0"},
+            },
+            {
+                {"Accept", "*/*"},
+            },
+            "",
+            [](Error error, Var<Response> response) {
+                REQUIRE(error);
```

Comment:
Here we should probably check for connection refused *or* timeout (guess that inside and outside polito things change significantly).